hybridse/src/udf/udf.cc

```diff
@@ -64,6 +64,39 @@ void hex(StringRef *str, StringRef *output) {
     output->data_ = buffer;
 }
 
+void unhex(StringRef *str, StringRef *output, bool* is_null) {
+    char *buffer = AllocManagedStringBuf(str->size_ / 2 + str->size_ % 2);
+    uint8_t* arr = new uint8_t[str->size_];
+    memset(arr, 0, str->size_);
+    for (uint32_t i=0; i < str->size_; i++) {
+        // only recognize 'A' to 'F' and 'a' to 'f' and '0' to '9', otherwise return null.
+        if (str->data_[i] <= 'F' && str->data_[i] >= 'A') {
+            arr[i] = str->data_[i] - 'A' + 10;
+        } else if (str->data_[i] <= 'f' && str->data_[i] >= 'a') {
+            arr[i] = str->data_[i] - 'a' + 10;
+        } else if (str->data_[i] <= '9' && str->data_[i] >= '0') {
+            arr[i] = str->data_[i] - '0';
+        } else {
+            *is_null = true;
+            break;
+        }
+    }
+    if (!*is_null) {    // every character is valid hex character
+        if (str->size_ % 2 == 0) {
+            for (uint32_t i=0; i < str->size_; i+=2) {
+                buffer[i/2] = static_cast<char>(arr[i] << 4 | arr[i+1]);
```

Comment:
rm the `arr` list and just access `str->data_`

Reply:
I use the `arr` list to store the decimal result of each character, and `(arr[i] << 4 | arr[i+1])` works when it's decimal, will it work when it's hex?

Reply:
no it won't work. But you can them into a single loop, as the logic `2char` -> `2 u8int ` -> `char`. Then arr is not needed